Src/Newtonsoft.Json.Tests/Utilities/StringUtilsTests.cs

```diff
@@ -50,14 +50,45 @@ public void ToCamelCaseTest()
             Assert.AreEqual("i", StringUtils.ToCamelCase("I"));
             Assert.AreEqual("", StringUtils.ToCamelCase(""));
             Assert.AreEqual(null, StringUtils.ToCamelCase(null));
-            Assert.AreEqual("iPhone", StringUtils.ToCamelCase("iPhone"));
             Assert.AreEqual("person", StringUtils.ToCamelCase("Person"));
+            Assert.AreEqual("iPhone", StringUtils.ToCamelCase("iPhone"));
             Assert.AreEqual("iPhone", StringUtils.ToCamelCase("IPhone"));
             Assert.AreEqual("i Phone", StringUtils.ToCamelCase("I Phone"));
+            Assert.AreEqual("i  Phone", StringUtils.ToCamelCase("I  Phone"));
             Assert.AreEqual(" IPhone", StringUtils.ToCamelCase(" IPhone"));
+            Assert.AreEqual(" IPhone ", StringUtils.ToCamelCase(" IPhone "));
             Assert.AreEqual("isCIA", StringUtils.ToCamelCase("IsCIA"));
             Assert.AreEqual("vmQ", StringUtils.ToCamelCase("VmQ"));
             Assert.AreEqual("xml2Json", StringUtils.ToCamelCase("Xml2Json"));
+            Assert.AreEqual("snAkEcAsE", StringUtils.ToCamelCase("SnAkEcAsE"));
+            Assert.AreEqual("snA__kEcAsE", StringUtils.ToCamelCase("SnA__kEcAsE"));
+            Assert.AreEqual("snA__ kEcAsE", StringUtils.ToCamelCase("SnA__ kEcAsE"));
+            Assert.AreEqual("already_snake_case_ ", StringUtils.ToCamelCase("already_snake_case_ "));
+        }
+
+        [Test]
+        public void ToSnakeCaseTest()
+        {
+            Assert.AreEqual("url_value", StringUtils.ToSnakeCase("URLValue"));
+            Assert.AreEqual("url", StringUtils.ToSnakeCase("URL"));
+            Assert.AreEqual("id", StringUtils.ToSnakeCase("ID"));
+            Assert.AreEqual("i", StringUtils.ToSnakeCase("I"));
+            Assert.AreEqual("", StringUtils.ToSnakeCase(""));
+            Assert.AreEqual(null, StringUtils.ToSnakeCase(null));
+            Assert.AreEqual("person", StringUtils.ToSnakeCase("Person"));
+            Assert.AreEqual("i_phone", StringUtils.ToSnakeCase("iPhone"));
+            Assert.AreEqual("i_phone", StringUtils.ToSnakeCase("IPhone"));
+            Assert.AreEqual("i_phone", StringUtils.ToSnakeCase("I Phone"));
+            Assert.AreEqual("i_phone", StringUtils.ToSnakeCase("I  Phone"));
+            Assert.AreEqual("i_phone", StringUtils.ToSnakeCase(" IPhone"));
+            Assert.AreEqual("i_phone", StringUtils.ToSnakeCase(" IPhone "));
+            Assert.AreEqual("is_cia", StringUtils.ToSnakeCase("IsCIA"));
```

Comment:
What is the expected output of something like IsJSONProperty? (capitalized acronym in middle of string) I assume it should be similar to how "URLValue" is resolved.